zaza/openstack/charm_tests/ceph/iscsi/tests.py

```diff
@@ -119,14 +146,67 @@ def mount_iscsi_target(self, ctxt):
             'iscsiadm --mode node --targetname {gw_iqn} --login']
         self.run_commands(ctxt['client_entity_id'], setup_cmds, ctxt)
 
-    def check_client_device(self, ctxt):
-        """Wait for multipath device to appear on client."""
+    def logout_iscsi_targets(self, ctxt):
+        """Logout of iscsi target on client.
+
+        :param ctxt: Base gateway context
+        :type ctxt: Dict
+        """
+        logging.info("Logging out of iscsi target")
+        logout_cmds = [
+            'iscsiadm --mode node --logoutall=all']
+        self.run_commands(ctxt['client_entity_id'], logout_cmds, ctxt)
+
+    def check_client_device(self, ctxt, init_client=True):
+        """Wait for multipath device to appear on client and test access.
+
+        :param ctxt: Base gateway context
+        :type ctxt: Dict
+        :param init_client: Initialise client if this is the first time it has
+                            been used.
+        :type init_client: bool
+        """
+        logging.info("Checking multipath device is present.")
+        device_ctxt = {
+            'bdevice': '/dev/dm-0',
+            'mount_point': '/mnt/iscsi',
+            'test_file': '/mnt/iscsi/test.data'}
+        ls_bdevice_cmd = 'ls -l {bdevice}'
+        mkfs_cmd = 'mke2fs {bdevice}'
+        mkdir_cmd = 'mkdir {mount_point}'
+        mount_cmd = 'mount {bdevice} {mount_point}'
+        umount_cmd = 'umount {mount_point}'
+        check_mounted_cmd = 'mountpoint {mount_point}'
+        write_cmd = 'truncate -s 1M {test_file}'
+        check_file = 'ls -l {test_file}'
+        if init_client:
+            commands = [
+                mkfs_cmd,
+                mkdir_cmd,
+                mount_cmd,
+                check_mounted_cmd,
+                write_cmd,
+                check_file,
+                umount_cmd]
```

Comment:
While the `mount -> write -> check -> unmount -> logout -> login -> mount -> check` test seems solid I would test the case where a share does not get umounted in the guest `mount -> write -> check -> logout -> login -> check`. In essence, that would validate that while the client might be blocked for some time, an interruption will be recoverable.